Sources/MongoSwiftSync/MongoCursor.swift

```diff
@@ -1,58 +1,126 @@
-     * the cursor will always be dead as soon as `next()` returns `nil`, or as soon as `nextOrError()` returns `nil` or
-     * throws an error.
+     * Indicates whether this cursor has the potential to return more data.
+     *
+     * This property is mainly useful if this cursor is tailable, since in that case `tryNext` may return more results
+     * even after returning `nil`.
+     *
+     * For non-tailable cursors, the cursor will always be dead as soon as `tryNext` returns `nil` or an error.
      */
-    public var isAlive: Bool { fatalError("unimplemented") }
+    public var isAlive: Bool {
+        return self.asyncCursor.isAlive
+    }
+
+    /// Returns the ID used by the server to track the cursor. `nil` once all results have been fetched from the server.
+    public var id: Int64? {
+        return self.asyncCursor.id
+    }
 
-    /// Returns the ID used by the server to track the cursor. `nil` until mongoc actually talks to the server by
-    /// iterating the cursor, and `nil` after mongoc has fetched all the results from the server.
-    public var id: Int64? { fatalError("unimplemented") }
+    /// Initializes a new `MongoCursor` instance from an async cursor. Not meant to be instantiated directly by a user.
+    internal init(wrapping cursor: MongoSwift.MongoCursor<T>, client: MongoClient) {
+        self.asyncCursor = cursor
+        self.client = client
+    }
 
     /**
-     * Initializes a new `MongoCursor` instance. Not meant to be instantiated directly by a user. When `forceIO` is
-     * true, this initializer will force a connection to the server if one is not already established.
+     * Returns a `Result` containing the next `T` in this cursor, an error if one occurred, or `nil` if the cursor is
+     * exhausted.
      *
-     * - Throws:
-     *   - `InvalidArgumentError` if the options passed to the command that generated this cursor formed an
-     *     invalid combination.
+     * If this cursor is tailable, this method will continue polling until a non-empty batch is returned from the server
+     * or the cursor is closed.
+     *
+     * - Returns:
+     *   A `Result<T, Error>?` containing the next `T` in this cursor on success, an error if one occurred, or `nil`
+     *   if the cursor was exhausted.
+     *
+     *   On failure, there error returned is likely one of the following:
+     *   - `CommandError` if an error occurs while fetching more results from the server.
+     *   - `LogicError` if this function is called after the cursor has died.
+     *   - `LogicError` if this function is called and the session associated with this cursor is inactive.
+     *   - `DecodingError` if an error occurs decoding the server's response.
      */
-    internal init(wrapping cursor: MongoSwift.MongoCursor<T>) throws {
-        fatalError("unimplemented")
+    public func next() -> Result<T, Error>? {
+        do {
+            return try self.asyncCursor.next().wait().map { .success($0) }
```

Comment:
non-awaitData tailable cursors have the potential to spam the server with a ton of requests and no real way for the user to rate limit them (we can't accept any arguments to `next` as part of the `Sequence` conformance). It would be possible for us to use `maxAwaitTimeMS` to configure the client side wait between `getNextDocument` calls, but that seems like a very nonstandard overload of the option. We could also introduce a new option that controls this on `find`/`aggregate`/others, but that seems needlessly confusing to users to introduce yet another `maxSomethingTimeMS`.

Reply:
I guess we do have an escape hatch, and given how infrequently pure tailable cursors are used, that may be enough.